lib/osutil/osutil.go

```diff
@@ -26,12 +26,54 @@ var renameLock = sync.NewMutex()
 // TryRename renames a file, leaving source file intact in case of failure.
 // Tries hard to succeed on various systems by temporarily tweaking directory
 // permissions and removing the destination file when necessary.
-func TryRename(filesystem fs.Filesystem, from, to string) error {
+func TryRename(from fs.Filesystem, fromPath string, to fs.Filesystem, toPath string) error {
 	renameLock.Lock()
 	defer renameLock.Unlock()
 
-	return withPreparedTarget(filesystem, from, to, func() error {
-		return filesystem.Rename(from, to)
+	return withPreparedTarget(from, fromPath, to, toPath, func() error {
+		// Optimization
+		if from == to || to == nil || (from.Type() == to.Type() && from.URI() == to.URI()) {
+			return from.Rename(fromPath, toPath)
+		}
+
+		// See if one is a prefix of the other filesystem.
+		if from.Type() == to.Type() {
+			fromUri := from.URI()
+			toUri := to.URI()
+			if strings.HasPrefix(toUri, fromUri) {
+				newToPath := filepath.Join(strings.TrimPrefix(toUri, fromUri), toPath)
+				if err := from.Rename(fromPath, newToPath); err == nil {
+					return err
+				}
+			} else if strings.HasPrefix(fromPath, toPath) {
+				newFromPath := filepath.Join(strings.TrimPrefix(fromUri, toUri), fromPath)
+				if err := to.Rename(newFromPath, toPath); err == nil {
+					return err
+				}
+			} else {
```

Comment:
The bigger issue is that I don't think we should pretend to provide something rename-like when in the background it's hundreds of network operations and syscalls and whatnot, where a failure leaves you in an unknown state with objects present on both sides, maybe fully copied maybe not, etc. It's not something we can successfully abstract. Instead, we should expose the fact that a cross-filesystem "rename" needs to be an open plus a create plus a copy plus a delete and let the user of the API decide what to do with failures at each of those points.

Except in the special case where an actual rename syscall can succeed.